src/platform/OpenThread/GenericThreadStackManagerImpl_OpenThread.cpp

```diff
@@ -992,6 +993,233 @@ CHIP_ERROR GenericThreadStackManagerImpl_OpenThread<ImplClass>::_JoinerStart(voi
+                memset(service, 0, sizeof(otSrpClientService));
+                service = next;
+            } while (service);
+        }
+        break;
+    }
+    case OT_ERROR_PARSE:
+        ChipLogError(DeviceLayer, "OnSrpClientNotification: Parsing operaton failed");
+        break;
+    case OT_ERROR_NOT_FOUND:
+        ChipLogError(DeviceLayer, "OnSrpClientNotification: Domain name or RRset does not exist");
+        break;
+    case OT_ERROR_NOT_IMPLEMENTED:
+        ChipLogError(DeviceLayer, "OnSrpClientNotification: Server does not support query type");
+        break;
+    case OT_ERROR_SECURITY:
+        ChipLogError(DeviceLayer, "OnSrpClientNotification: Operation refused for security reasons");
+        break;
+    case OT_ERROR_DUPLICATED:
+        ChipLogError(DeviceLayer, "OnSrpClientNotification: Domain name or RRset is duplicated");
+        break;
+    case OT_ERROR_RESPONSE_TIMEOUT:
+        ChipLogError(DeviceLayer, "OnSrpClientNotification: Timed out waiting on server response");
+        break;
+    case OT_ERROR_INVALID_ARGS:
+        ChipLogError(DeviceLayer, "OnSrpClientNotification: Invalid service structure detected");
+        break;
+    case OT_ERROR_NO_BUFS:
+        ChipLogError(DeviceLayer, "OnSrpClientNotification: Insufficient buffer to handle message");
+        break;
+    case OT_ERROR_FAILED:
+        ChipLogError(DeviceLayer, "OnSrpClientNotification: Internal server error occurred");
+        break;
+    default:
+        ChipLogError(DeviceLayer, "OnSrpClientNotification: Unknown error occurred");
+        break;
+    }
+}
+
+template <class ImplClass>
+void GenericThreadStackManagerImpl_OpenThread<ImplClass>::_InitSrpClient()
+{
+    otSrpClientSetCallback(mOTInst, &OnSrpClientNotification, nullptr);
+}
+
+template <class ImplClass>
+CHIP_ERROR GenericThreadStackManagerImpl_OpenThread<ImplClass>::_AddSrpService(const char * aInstanceName, const char * aName,
+                                                                               uint16_t aPort, uint32_t aLeaseInterval,
+                                                                               uint32_t aKeyLeaseInterval)
+{
+    otError error                   = OT_ERROR_NONE;
+    otSrpClientService * srpService = nullptr;
+
+    VerifyOrExit(aInstanceName, error = OT_ERROR_INVALID_ARGS);
+    VerifyOrExit(aName, error = OT_ERROR_INVALID_ARGS);
+
+    // Check if service with desired instance name already exists.
+    for (otSrpClientService & service : mSrpServices)
+    {
+        if (service.mInstanceName)
+        {
+            VerifyOrExit((strcmp(service.mInstanceName, aInstanceName) != 0), error = OT_ERROR_DUPLICATED);
+        }
+        else
+        {
+            srpService = srpService ? srpService : &service;
+        }
+    }
+
+    // Verify is there an empty place for new service.
+    VerifyOrExit(srpService, error = OT_ERROR_NO_BUFS);
+
+    memset(srpService, 0, sizeof(otSrpClientService));
```

Comment:
And what if we call AddSrpService for the first time, so srpService will point to the first element of array, which has never been used before? I guess it might be filled with trash.

Reply:
I think you should zero the array in the initialization function. Otherwise, you may never find a service which has never been used ;)